add e2e docs (#7512)
RasaHQ/rasa: docs/docs/policies.mdx

This policy has a pre-defined architecture, which comprises the

> following steps:

Reviewer: I think I'm missing some high-level overview of the different things that TED does. Could we add such thing before enumerating these steps? From the detailed steps below I get the idea that TED 1) predicts next action based on the starspace-inspired similarity (though this isn't said explicitly), 2) predicts entities (in the most recent user turn?), but it requires a lot of effort to extract this high-level info from the steps.

Author: copied description form DIET

> 1. Concatenate user input (user intent and entities) or user text processed through a user sequence transformer encoder,
>    previous system actions or bot utterances processed through a bot sequence transformer encoder,
>    slots and active forms for each time
>    step into an input vector to pre dialogue transformer embedding layer.

Reviewer: ```suggestion
1. Concatenate features for
   - user input (user intent and entities) or user text processed through a user sequence transformer encoder,
   - previous system actions or bot utterances processed through a bot sequence transformer encoder,
   - slots and active forms

   for each time step into an input vector to the embedding layer that precedes the 
   dialogue transformer.
```

> 2. Feed the input vector into a dialogue transformer encoder.

Reviewer: ```suggestion
2. Feed the embedding of the input vector into the dialogue transformer encoder.
```

> 3. Apply a dense layer to the output of the dialogue transformer to get embeddings of a dialogue for each time step.

Reviewer: ```suggestion
3. Apply a dense layer to the output of the dialogue transformer to get embeddings of the dialogue for each time step.
```

4. Apply a dense layer to create embeddings for system actions for each time step.

5. Calculate the similarity between the dialogue embedding and embedded system actions.
   This step is based on the [StarSpace](https://arxiv.org/abs/1709.03856) idea.

> 6. Concatenate the token output of the user sequence transformer encoder

Reviewer: ```suggestion
6. Concatenate the token-level output of the user sequence transformer encoder
```

  Also, it is usual practice to have decreasing values in the list: next value is smaller or equal to the
  value before.

> * `sequence_number_of_transformer_layers`:

Reviewer 2: I am confused. This does not match the with the changes you mentioned earlier or am I missing something?

> number_of_transformer_layers -> dictionary “number_of_transformer_layers” with keys “text”, “action_text”, “label_action_text”, “dialogue”

Reviewer 2: same for the others

Reviewer 2: Ok, I think you simply forgot to update these here.

Author: yep simply forgot to remove

<details><summary>More configurable parameters</summary>

```
+---------------------------------------+------------------------+--------------------------------------------------------------+
| Parameter                             | Default Value          | Description                                                  |
+=======================================+========================+==============================================================+
| hidden_layers_sizes                   | text: []               | Hidden layer sizes for layers before the embedding layers    |
|                                       | action_text: []        | for user messages and bot messages in previous actions       |
|                                       | label_action_text: []  | and labels. The number of hidden layers is                   |
|                                       |                        | equal to the length of the corresponding list.               |
+---------------------------------------+------------------------+--------------------------------------------------------------+

> | dense_dimension                       | text: 128              | Dense dimension for sparse features to use.                  |
> |                                       | action_text: 128       |                                                              |

Reviewer: ```suggestion
| dense_dimension                       | text: 128              | Dense dimension for sparse features to use after they are    |
|                                       | action_text: 128       | converted into dense features.                               |
```

|                                       | label_action_text: 128 |                                                              |
|                                       | intent: 20             |                                                              |
|                                       | action_name: 20        |                                                              |
|                                       | label_action_name: 20  |                                                              |
|                                       | entities: 20           |                                                              |
|                                       | slots: 20              |                                                              |
|                                       | active_loop: 20        |                                                              |
+---------------------------------------+------------------------+--------------------------------------------------------------+

> | concat_dimension                      | text: 128              | Concat dimension for sequence and sentence features.         |
> |                                       | action_text: 128       |                                                              |

Reviewer: ```suggestion
| concat_dimension                      | text: 128              | Common dimension to which sequence and sentence features of  |
|                                       | action_text: 128       | different dimensions get converted before concatenation.     |
```

|                                       | action_text: 128       | Number of units in bot text sequence transformer encoder.    |
|                                       | label_action_text: 128 | Number of units in bot text sequence transformer encoder.    |
|                                       | dialogue: 128          | Number of units in dialogue transformer encoder.             |
+---------------------------------------+------------------------+--------------------------------------------------------------+
| number_of_transformer_layers          | text: 1                | Number of layers in user text sequence transformer encoder.  |
|                                       | action_text: 1         | Number of layers in bot text sequence transformer encoder.   |
|                                       | label_action_text: 1   | Number of layers in bot text sequence transformer encoder.   |
|                                       | dialogue: 1            | Number of layers in dialogue transformer encoder.            |
+---------------------------------------+------------------------+--------------------------------------------------------------+

> | number_of_attention_heads             | 4                      | Number of attention heads in transformers.                   |

Reviewer: ```suggestion
| number_of_attention_heads             | 4                      | Number of self-attention heads in transformers.                   |
```

|                                       |                        | If constant `batch_size` is required, pass an int, e.g. `8`. |
+---------------------------------------+------------------------+--------------------------------------------------------------+
| batch_strategy                        | "balanced"             | Strategy used when creating batches.                         |
|                                       |                        | Can be either 'sequence' or 'balanced'.                      |
+---------------------------------------+------------------------+--------------------------------------------------------------+
| epochs                                | 1                      | Number of epochs to train.                                   |
+---------------------------------------+------------------------+--------------------------------------------------------------+
| random_seed                           | None                   | Set random seed to any 'int' to get reproducible results.    |
+---------------------------------------+------------------------+--------------------------------------------------------------+

> | embedding_dimension                   | 20                     | Dimension size of embedding vectors.                         |

Reviewer: ```suggestion
| embedding_dimension                   | 20                     | Dimension size of dialogue & system action embedding vectors.|
```

Reviewer: Meant to better disambiguate `embedding_dimension` and `encoding_dimension`.

+---------------------------------------+------------------------+--------------------------------------------------------------+
| drop_rate_attention                   | 0.0                    | Dropout rate for attention. Value should be between 0 and 1. |
|                                       |                        | The higher the value the higher the regularization effect.   |
+---------------------------------------+------------------------+--------------------------------------------------------------+
| weight_sparsity                       | 0.8                    | Sparsity of the weights in dense layers.                     |
|                                       |                        | Value should be between 0 and 1.                             |
+---------------------------------------+------------------------+--------------------------------------------------------------+
| use_sparse_input_dropout              | True                   | If 'True' apply dropout to sparse input tensors.             |
+---------------------------------------+------------------------+--------------------------------------------------------------+

> | use_dense_input_dropout               | True                   | If 'True' apply dropout to dense input tensors.              |

Reviewer: I think this isn't accurate: it describes dropout applied to _sparse_ input tensors after they are converted into dense ones...

Author: good catch

|                                       |                        | logged. Either after every epoch ('epoch') or for every      |
|                                       |                        | training step ('minibatch').                                 |
+---------------------------------------+------------------------+--------------------------------------------------------------+
| checkpoint_model                      | False                  | Save the best performing model during training. Models are   |
|                                       |                        | stored to the location specified by `--out`. Only the one    |
|                                       |                        | best model will be saved.                                    |
|                                       |                        | Requires `evaluate_on_number_of_examples > 0` and            |
|                                       |                        | `evaluate_every_number_of_epochs > 0`                        |
+---------------------------------------+------------------------+--------------------------------------------------------------+

> | e2e_confidence_threshold              | 0.5                    | The threshold that ensures that end-to-end is picked only if |

Reviewer: What is a reasonable value range for this one?

Author: I have no idea. The logic behind `0.5` is that in this case we're sure that it is winning label

Reviewer: Okay then. I reckon that with e2e there will be things where we simply don't have good advice yet 🙂

Author: we don't even have the bad one 😅
